### PR TITLE
add support for --message-format option in `cargo fmt`

### LINKED ISSUE #3274: `cargo fmt` doesn't accept `--message-format`.
This means it can't easily be invoked by tools (e.g. editors) that want to avoid dealing with multiline error messages.

#2817 is related, in that `cargo fmt` should probably accept most general Cargo flags.

cc @Manishearth @oli-obk Does `cargo clippy` do anything special here?

### PR DESCRIPTION
**Pull request by calebcartwright**
Closes #3274

`--message-format` in the other `cargo` subcommands accepts 3 values:

1. human
2. short
3. json

In the context of rustfmt, my thinking was:

1. human -> basically a noop, defaults to the standard/regular output
2. short -> list the names of files with diffs
3. json -> maps to the `json` emit mode

IIRC, using `--message-format short` with `clippy` lists the file names along with the line numbers with lint violations, but I didn't think trying to include line numbers made as much sense for rustfmt in the `short` mode.

Each of these output/message formats is already available via other flags/options that have been recently added (`--emit json`, `-l`, `--files-with-diff`), so the implementation is basically just mapping the `--message-format` value to those corresponding args rustfmt already supports.

As examples:
`short`
```bash
$ cargo +nightly fmt --message-format short -- --check
src/lib.rs
src/foo.rs
```
`json` (pretty printed after for ease of view)
```bash
$ cargo +nightly fmt --message-format json
[  
  {  
    "name":"tests/writemode/source/json.rs",
    "mismatches":[  
      {  
        "original_begin_line":5,
        "original_end_line":7,
        "expected_begin_line":5,
        "expected_end_line":5,
        "original":"fn foo_expr() {\n    1\n}",
        "expected":"fn foo_expr() { 1 }"
      },
      {  
        "original_begin_line":9,
        "original_end_line":11,
        "expected_begin_line":7,
        "expected_end_line":7,
        "original":"fn foo_stmt() {\n    foo();\n}",
        "expected":"fn foo_stmt() { foo(); }"
      },
      {  
        "original_begin_line":13,
        "original_end_line":15,
        "expected_begin_line":9,
        "expected_end_line":9,
        "original":"fn foo_decl_local()  {\n    let z = 5;\n   }",
        "expected":"fn foo_decl_local() { let z = 5; }"
      },
      {  
        "original_begin_line":17,
        "original_end_line":19,
        "expected_begin_line":11,
        "expected_end_line":11,
        "original":"fn    foo_decl_item(x: &mut i32) {\n    x = 3;\n}",
        "expected":"fn foo_decl_item(x: &mut i32) { x = 3; }"
      },
      {  
        "original_begin_line":21,
        "original_end_line":21,
        "expected_begin_line":13,
        "expected_end_line":13,
        "original":"   fn empty()     {",
        "expected":"fn empty() {}"
      },
      {  
        "original_begin_line":23,
        "original_end_line":23,
        "expected_begin_line":15,
        "expected_end_line":15,
        "original":"}",
        "expected":"fn foo_return() -> String { \"yay\" }"
      },
      {  
        "original_begin_line":25,
        "original_end_line":29,
        "expected_begin_line":17,
        "expected_end_line":20,
        "original":"fn foo_return() -> String {\n    \"yay\"\n}\n\nfn foo_where() -> T where T: Sync {",
        "expected":"fn foo_where() -> T\nwhere\n    T: Sync,\n{"
      },
      {  
        "original_begin_line":64,
        "original_end_line":66,
        "expected_begin_line":55,
        "expected_end_line":55,
        "original":"fn lots_of_space                      ()                                                           {\n                           1                 \n}",
        "expected":"fn lots_of_space() { 1 }"
      },
      {  
        "original_begin_line":71,
        "original_end_line":72,
        "expected_begin_line":60,
        "expected_end_line":60,
        "original":"    fn dummy(&self) {\n    }",
        "expected":"    fn dummy(&self) {}"
      },
      {  
        "original_begin_line":75,
        "original_end_line":75,
        "expected_begin_line":63,
        "expected_end_line":64,
        "original":"trait CoolerTypes { fn dummy(&self) { ",
        "expected":"trait CoolerTypes {\n    fn dummy(&self) {}"
      },
      {  
        "original_begin_line":77,
        "original_end_line":77,
        "expected_begin_line":66,
        "expected_end_line":66,
        "original":"}",
        "expected":""
      },
      {  
        "original_begin_line":79,
        "original_end_line":79,
        "expected_begin_line":67,
        "expected_end_line":70,
        "original":"fn Foo<T>() where T: Bar {",
        "expected":"fn Foo<T>()\nwhere\n    T: Bar,\n{"
      }
    ]
  }
]
```

### CODE REVIEW

File: src/cargo-fmt/main.rs [hunk 1]
```diff
@@ -100,6 +104,14 @@ fn execute() -> i32 {
     }
 
     let strategy = CargoFmtStrategy::from_opts(&opts);
+    let mut rustfmt_args = opts.rustfmt_options;
+    if let Some(message_format) = opts.message_format {
+        if let Err(msg) = convert_message_format_to_rustfmt_args(message_format, &mut rustfmt_args)
+        {
+            print_usage_to_stderr(&msg);
+            return FAILURE;
+        }
```
Comment (calebcartwright) [inline src/cargo-fmt/main.rs]: I started out implementing everything inline here, but extracted it out into a separate function to make it easier to test

File: src/cargo-fmt/main.rs [hunk 2]
```diff
@@ -110,16 +122,53 @@ fn execute() -> i32 {
...
-            &strategy,
-            opts.rustfmt_options,
-            None,
-        ))
+        handle_command_status(format_crate(verbosity, &strategy, rustfmt_args, None))
+    }
+}
+
+fn convert_message_format_to_rustfmt_args(
+    message_format: String,
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: Would you mind changing this from `String` to `&str`? I think that in Rust it is usually better to use `&str` over `String` unless you are modifying or acquiring the ownership of that string.

File: src/cargo-fmt/main.rs [hunk 3]
```diff
@@ -110,16 +122,53 @@ fn execute() -> i32 {
...
+    rustfmt_args: &mut Vec<String>,
+) -> Result<(), String> {
+    let contains_emit_mode = rustfmt_args.contains(&String::from("--emit"));
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: We do not need to create a `String` and then take its reference, just passing the string literal will do :)

Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: That's what I was originally trying to do, but the compiler was not having it. Combination of `rustfmt_args` containing `String`s and what `contains` wanted 🤷‍♂ 

I'll have another go at it

File: src/cargo-fmt/main.rs (hunk 2 above)
Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: Done!

File: src/cargo-fmt/main.rs (hunk 3 above)
Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: Btw I believe https://github.com/rust-lang/rust/issues/42671 is the same issue I was running into while trying to use the string literal.

I switched this to use the workaround described on that thread

Comment (topecongiro) [reply src/cargo-fmt/main.rs]: Ah, I have forgotten about this...I remember hitting this a while ago myself. Sorry!

**Comment by topecongiro** [comment]
Thanks!

**Pull request merged**